Create context class

stadelmanma opened the pull request:
This creates a Context class instance that inherits from the Hash core class allowing most of the old interface to remain unchanged. 

A single Context instance is created per template by moving it from Document.process into Template.render. The current template instance is registered to the Context class instance to allow for any future back references that may come in handy. A `nil` value can be passed through in the absence of a template (i.e. unit tests). It seems for all practical cases (outside of unit testing), a context instance is useless without a template. 

This PR lays the foundation to address #43

senny commented:
I am not sure about the proposed changes. The context to me is a thing you render the template with. It's not including the template itself. It's possible that a new object helps that groups a template and a context together. I'm especially worried to the changes on `Sablon::Content.wrap`. This is public interface and also used by components in the stack I was authoring.

Another train of thought, maybe we need to refactor towards passing a content more information when it is applied. Meaning not during instantiation but when `append_to` is called. This would be a "late binding" approach and could yield some additional benefits. What do you think?

stadelmanma commented:
I completely agree with you in regards to the context generated by the user and passed into Sablon is the thing you render the template with and not tied to the template itself. Perhaps a new name for the context + template object would better clarify that. Wrapping the two together behind the scenes allows for more flexibility in the future when looking towards more complex cases, i.e. pulling content from a relative location instead of being forced to wrap it early on. 

The main issue we will run into with passing the context late is that when processing HTML, any lists are registered with the Numbering singleton during initialization. Looking at how the code is structured I think it would be very difficult to store the given list definitions somewhere temporarily then magically set the id's once the entire WordML syntax tree is known at the `append_to` stage. To get around that issue it will likely be much easier to defer parsing of the HTML code until `append_to` is called instead of during initialization.

senny commented:
> To get around that issue it will likely be much easier to defer parsing of the HTML code until append_to is called instead of during initialization.

Yes, that would need to happen if we decide to go down the late binding road. I can't remember from the top of my head if `document.xml` is processed before the numbering definitions are written. If that is the case, it should continue to work if we move:

```ruby
word_ml = Sablon.content(:word_ml, converter.process(html))
```

Which would need to happen anyway as `process()` will require a template context to be passed in, that won't be available beforehand.

Do you think the template + context grouping is still necessary even if we do pass more contextual data to `append_to`?

stadelmanma commented:
`document.xml` is processed before `numbering.xml`, the files in the Zip are sorted by name.

The template relationship with context could be dropped but you will need need a Context instance to store the Numbering instance on. It is reasonable to store a numbering instance there because it is context dependent when lists are being generated.

I'm not sure where you want that block to be placed? My changes to the HTML portion of content.rb are shown below in my [rework numbering branch](https://github.com/stadelmanma/sablon/tree/remove-singleton-behavior-from-numbering). They should be equivalent and it keeps your external, public API from being mixed in with the internals.
```ruby
    # Handles conversion of HTML -> WordML and addition into template
    class HTML < Struct.new(:html_content)
      include Sablon::Content
      def self.id
        :html
      end

      def self.wraps?(_)
        false
      end

      def initialize(value)
        super value
      end

      def append_to(paragraph, display_node, context)
        converter = HTMLConverter.new(context)
        word_ml = WordML.new(converter.process(@html_content))
        word_ml.append_to(paragraph, display_node, context)
      end
    end
```

senny commented:
@stadelmanma I think we pretty much agree on the direction. I don't like the naming collision between "the context a user passes to render sablon" and the new `Context` object. I also don't think that new object needs to subclass `Hash`. Maybe `Environment` could be used as a name for the object that holds the user passed `context` and provides an API to store processing specific data like the numbering definitions.

The snippet you pasted look fine. Maybe the converter does not need the `environment` upon initialization but we can pass it to `process`. Would allow for some re-use of the converter object but doesn't matter much.

stadelmanma commented:
Relevant methods from the converter. I chose to store it as an instance variable to avoid modifying several methods just to pass the context down the call chain. 

```ruby
    def initialize
      @context = nil
      @numbering = nil
    end

...

    def process(input, context)
      @context = context
      @numbering = @context.numbering
      processed_ast(input).to_docx
    end
```

senny commented:
@stadelmanma I don't have the full picture but maybe it's worth to make the code explicit and unpack the `context` from the new `env`. Meaning if we start to pass around a `env` that is an aggregate of the `context`, `template` and some intermediate storage for numbering etc. Calling code that want's the context can be explicit about it:

```ruby
env.context[:var]
```

concrete the `Inseration` class would look like this:

```ruby
      def evaluate(env)
        if content = expr.evaluate(env.context)
          field.replace(Sablon::Content.wrap(content))
        else
          field.remove
        end
      end
```

maybe `field.replace` will need the `env` as well to later pass it to `append_to`.

While this will be a lot of renaming it allows us to keep the current understanding of what a `context` is and add the new meaning of the `environment`.

stadelmanma commented:
I'm all for the renaming but after some more through work and debugging the way blocks are handled throws a wrench in being completely explicit with handling of the environment. 

Excerpt from Statement::Loop 
```ruby
        content = value.flat_map do |item|
          iteration_context = env.context.merge(iterator_name => item)
          block.process(iteration_context)
        end
```

Since any arbitrary block can get processed you would have to instantiate a new env every iteration and then clone over the attributes.

senny commented:
@stadelmanma is it problematic to instantiate a new environment? I mean it's currently also happening with the `.merge` that we create a new context for each iteration. Maybe a method on `Environment` could help here:

```ruby
class Environment
  def expand(expanded_context)
    Environment.new(context.merge(expanded_context), template, other)
  end
end
```

stadelmanma commented:
@senny, so I have the code to remove the singleton behavior from Numbering mostly ready, built upon this PR. However, I am getting very odd behavior when in the html_test.rb file. When I run either one of the tests on their own it will always perform as expected (passing). 

However, when I run either `bundle exec rake` or `bundle exec ruby -I"lib:test" test/html_test.rb --verbose` either one or both will fail. After doing some debugging it appears that for the html_test.rb file the environment and numbering instance variables are getting reused even through they get recreated by `Environment.new` in `Template.render`.

```
Environment Inititalized! Env: #<Sablon::Environment:0x007f8048219f10>Num: #<Sablon::Numbering:0x007f8048219ee8>
In HTMLConverter.process Env: #<Sablon::Environment:0x007f8048219f10>Num: #<Sablon::Numbering:0x007f8048219ee8>
SablonHTMLTest#test_generate_document_from_template_with_styles_and_html = 0.07 s = .

Enivonment Inititalized! Env: #<Sablon::Environment:0x007f80498390e8>Num: #<Sablon::Numbering:0x007f8049838ff8>
In HTMLConverter.process Env: #<Sablon::Environment:0x007f8048219f10>Num: #<Sablon::Numbering:0x007f8048219ee8>
SablonHTMLTest#test_generate_document_from_template_without_styles_and_html = 0.05 s = F
```

Any insight as to why my instances are getting mixed up? This only affects the code in my my [rework numbering branch](https://github.com/stadelmanma/sablon/tree/remove-singleton-behavior-from-numbering).

stadelmanma commented:
So moving my usage of an environment out of the parser, stored as instance variable on the parser, and into field.replace fixed the issue. https://github.com/stadelmanma/sablon/commit/bcde06fa689af7c1212c073c11369b84e3aca240

Not entirely sure why that fixed the problem since a new parser should have been created between the two tests in html_test.rb.

senny commented on lib/sablon/environment.rb:
```diff
@@ -0,0 +1,47 @@
+      @context = transform_hash(context)
+    end
+
+    def transform_hash(hash)
```
Could we keep those functions in the `Context` module to make it explicit that they are about the context not the environment?

senny commented on lib/sablon/operations.rb:
```diff
@@ -12,24 +12,24 @@ def evaluate(context)
         content = value.flat_map do |item|
-          iteration_context = context.merge(iterator_name => item)
-          block.process(iteration_context)
+          iter_env = env.alter_context({iterator_name => item})
```
The `{` and `}` can be omitted.

senny commented on test/sandbox/environment_test.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  def test_does_not_wrap_html_and_wordml_with_nil_value
+    env = Sablon::Environment.new(nil, {"html:mykey" => nil, "word_ml:otherkey" => nil, "normalkey" => nil})
```
The `{` and `}` can be omitted if it's the last argument of a method call. This applies to all previous tests as well.

stadelmanma commented:
I've added your changes I also added some better testing of the environment class since I realized I only actually tested the internal context object.

senny commented:
I ended up extracting the `Context` module into a separate file again. 1c39b0312493de30b88e8de4b9de6698a84fdf9a
